### PR TITLE
(GH-1964) Add deprecation warning utility

### LINKED ISSUE #1964: Standardize Deprecation warnings
## Use Case
Bolt is moving rapidly and deprecates a feature every month or so we want to 

## Describe the Solution You Would Like
A standard function for deprecation warnings so that we ensure consistency and add usability features like silencing some warnings.

- This function should wrap "warn_once"
- This function should report an analytics event with a unique deprecation id for each feature.
- This function should allow the issueing of silent deprecations that only send analytics.

## Additional Context
When should we start sending warnings.
- Plan language deprecations should wait until near the major version release so people don't have to update their contents.
- Smaller deprecation should have ASAP since people don't necessarily upgrade frequently.
- Most deprecation warnings aren't seen by most users or are simple. (ie debug -> loglevel)
- experimental features should change as fast as possible so we can validate our ideas.

### PR DESCRIPTION
**Pull request by lucywyman**
Previously, we had no way to track how many users were experiencing
warnings for deprecated features, or which deprecated features were
being used. This commit adds a new class method `deprecation_warning` to
the `Bolt::Logger` class, which sends an analytics event with which type
of warning was issued then calls `warn_once` for the deprecation. This
also changes existing deprecation warnings to use the new deprecation
method, and audits existing warnings to determine if they should use
`warn_once` or `deprecation_warning`.

Closes #1964

!no-release-note

### CODE REVIEW

File: resources/bolt_bash_completion.sh [hunk 1]
```diff
@@ -6,74 +6,134 @@
 ###
 ### Marc Schoechlin ms-github@256bit.org
 
-_bolt()
```
Comment (beechtom) [inline resources/bolt_bash_completion.sh:9]: I think you need a rebase since this looks like work from another PR?

Comment (lucywyman) [reply resources/bolt_bash_completion.sh]: I think this is just cause it's against main instead of master. Updated

File: lib/bolt/logger.rb [hunk 2]
```diff
@@ -67,6 +68,10 @@ def self.configure(destinations, color)
       end
     end
 
+    def self.analytics(analytics = Bolt::Analytics::NoopClient.new)
```
Comment (nicklewis) [inline lib/bolt/logger.rb:71]: I'm not clear what this method is supposed to do. Is it a getter or a setter or both?

Comment (lucywyman) [reply lib/bolt/logger.rb:71]: It's meant to just be a setter, but calling it `set_analytics` makes rubocop unhappy. There's probably a better way.

Comment (nicklewis) [reply lib/bolt/logger.rb:71]: In that case, it should be `def self.analytics=(analytics)`. No default value for the argument then.

File: lib/bolt/logger.rb [hunk 3]
```diff
@@ -67,6 +67,10 @@ def self.configure(destinations, color)
       end
     end
 
+    def self.analytics=(analytics)
+      @analytics ||= analytics
```
Comment (nicklewis) [inline lib/bolt/logger.rb:71]: This should just be straight assignment rather than `||=`.

File: lib/bolt/cli.rb [hunk 4]
```diff
@@ -131,6 +131,7 @@ def parse
                   end
 
         Bolt::Logger.configure(config.log, config.color)
+        Bolt::Logger.analytics(analytics)
```
Comment (nicklewis) [inline lib/bolt/cli.rb:134]: This needs to be updated to use assignment.

Comment (lucywyman) [reply lib/bolt/cli.rb:134]: Oh no sorry I know still working on this - I pushed too soon.

File: bolt-modules/boltlib/lib/puppet/datatypes/target.rb [hunk 5]
```diff
@@ -1,5 +1,8 @@
 # frozen_string_literal: true
 
+# This was failing local test runs
+require 'bolt/target'
```
Comment (nicklewis) [inline bolt-modules/boltlib/lib/puppet/datatypes/target.rb:4]: We should fix the tests that aren't requiring this properly. The target class should come from the inventory which is injected via the puppet context. This type doesn't depend on `Bolt::Target`, which is important for use in PE.

Comment (lucywyman) [reply bolt-modules/boltlib/lib/puppet/datatypes/target.rb]: I don't remember which test was failing - I'll just remove this for now.

**Pull request merged**